### DISCUSSION TITLE
How to undo file deletion?

### DISCUSSION THREAD

**Discussion by egor7orlov** [zed-industries/zed#9143] [Feature Requests]
How to undo file deletion?

`cmd+z` doesn't seem to work. And there is no `Undo` option in the context menu (when I click RMB on the file tree or Edit tab above the editor).

**Comment by Moshyfawn** [comment]
It hasn't been implemented in Zed yet. You can find the corresponding feature request in #5039. Be sure to +👍.

**Comment by 6lackknight** [comment]
I love Zed, but really??

I only now realising that this is a feature that doesn't exist. I've just lost about an hour of changes because I selected the wrong files for deletion.

**Comment by 1hko** [comment]
Also no Undo File Rename. If renaming would automatically update imports in other files, I would be panicked by inability to undo the change.

**Comment by devas-37** [comment]
Блин я только что удалили файл с изменением два дня. Удаляю зед...*?;%("%!?":%?*№("%:

**Comment by oleggromov** [comment]
+1 this. This is a terrible idea to provide a "Trash" and "Delete" actions side by side without the way of undoing the last one. Just posting out of frustration - thankfully, there was only a handful of files I accidentally deleted, not days worth of work.

**Comment by kettle11** [comment]
I just lost probably 8 hours of focused work due to this. I never thought my text editor would have an irreversible delete option so easily available and I was tidying up some code before making a commit so I renamed a file then deleted an outdated file, but whoops I had selected delete the new file by accident.

No problem. I'll just Ctrl+Z. Ah, that's not a feature in Zed. No problem I'll just check the trash, but nope the 'Delete' feature skips the trash. Well I'll get it with git, ah that's right I just renamed the file. Well I'll use Time Machine, but that's right this is a new Mac and I forgot to set that up yet.

So now I'm running a file disk scan with some tool that tries to recover files from the filesystem, but I don't expect it to work.

This design in Zed is atrocious! It's a massive footgun cocked and loaded waiting for people as unlucky as myself to delete tons of their work.

**Reply by tyrael147** [reply]
I do not know if this is useful for you but I had the exact problem some hours ago in Ubuntu 24 and Zed v0.215.3

In my home directory I searched for the file I deleted by accident (hoping to find some *.pyc or something), originally named "main_batch_task.py" by doing `find . -name "*main_batch*"`.

I got this result:

```
./.local/share/Trash/files/main_batch_task.py
./.local/share/Trash/info/main_batch_task.py.trashinfo
```
I then cp'ed the file into my repo and I verified that it was indeed the file I deleted. I am a rookie Linux user so I do not know if this behavior is due to Linux or Zed.

I hope this helps.

Cheers.

**Reply by big-yellow-duck** [reply]
can confirm this saved me in fedora linux

**Comment by MJKhaani** [comment]
Hi there.

It's extremely frustrating that because the mouse right click does not change selected file, the code that i worked for 1 week got deleted. I got the last commit, but the one week of changes is lost.

@ tyrael147 way is not working on macOS.

I loved Zed the first time that i used it, but i considering to switch that ugly VSCode, at least this would not happen again.

Thanks

**Comment by chooisfox** [comment]
Yeah, just got fucked by zed just like all other previous people. Click didn’t register and I’ve just deleted almost a week worth of work. Can’t restore it, cause .git folder got deleted as well :/

**Reply by chooisfox** [reply]
Worst part that this was an issue almost 2 years ago

**Comment by JonasOstir** [comment]
One could always pray that the `Pane: reopen closed item` works 🤞🏻. That is how I could restore a file that I accidentally removed.

**Reply by melgio** [reply]
this lowkey saved me hours of time, still amazed that there is no dedicated undo to file delete

**Comment by markvh033** [comment]
Im sorry but what is this???? I accidentally delete an item and its just gone forever? Cannot work like that. Feels like were 10 years back in time again. Back to VScode

**Comment by michaelcamper** [comment]
This is incredibly frustrating

**Comment by jpmonette** [comment]
Just joined the club - been working for the past 5 hours on a project, and instead if deleting 2 files, I deleted all but those 2 files. 🥲

**Comment by BjornFridal** [comment]
I have been using Zed for a week. It's a little lacking when it comes to extensions, but the speed makes up for it. Nice and simple! However I've just accidentally deleted the wrong file. I'll be back once this is fixed :)

**Comment by joakim** [comment]
This is a serious bug IMO. In no app should it be acceptable that the user permantently and irreversably loses their data/work. It blows my mind that this is the norm in some IDEs.

**Comment by heycuauh** [comment]
I’ve finally joined the club. I’m not even sure what I did. I randomly hit a  key combination while switching between virtual desktops, and bam! One hour of work gone.

**Comment by joshistoast** [comment]
I right clicked a folder I wanted to undo, then the entire project got undone :(

**Reply by egor7orlov** [reply]
It was a lesson for me not to jump into new hot technology and rely only on the most popular and boring stuff (VSCode)

**Comment by softwareentrepreneer** [comment]
still waiting...

**Reply by softwareentrepreneer** [reply]
I always need to pay extra attention on this issue.
For example my project will download some data for testing, whenever I want to remove the obsolete data (I swear I already clicked the folder...), sometimes I just accidentally removed my code file...and I have to check my git changes, manually and carefully move it back to the repo....This is **REALLY REALLY ANNOYING**
As a new user (only been using zed for < 1 month), this issue and the lack of extensions have been my only complaints so far. everything else is pretty good.

**Comment by itsisaac19** [comment]
still waiting on this

**Comment by 6lackknight** [comment]
This should not be closed @egor7orlov @Moshyfawn -- this is an active issue and has been for years.

Are there any plans / current work on this?

**Reply by egor7orlov** [reply]
Yeah, maybe, since [feature](https://github.com/zed-industries/zed/issues/5039) that was mentioned in the first reply is now closed for some reason.
I ditched Zed, so I don't really care anymore🥱

**Comment by slipperystairs** [comment]
Please, I just deleted my entire file system.

**Reply by vwkd** [reply]
Run `rm -rf /`. It might fix things.

**Comment by chrisveness** [comment]
I did Ctrl-Z to undo my last typing, but focus was apparently in the Project Panel, not the editor.

Ctrl-Z undid the file copy for the file I've been working on for the last two or three hours – apparently the file is deleted, with no warning, and no way to recover it. WHAT!!

**Reply by 6lackknight** [reply]
can't believe this is still a thing.

maybe we need to all band together and fix this once and for all.
